### PR TITLE
user provided bound for torchtrt compile when export dimension is unb…

### PR DESCRIPTION
**Pull request by apbose**
This PR  validates user bounds against any finite exporter bounds, with three branches:                                                                                                                                                                                                                                                                                                                                     
  - Upper overflow (user_max > exp_max) → ValueError. TRT engine profile follows the exporter; shapes above exp_max would be rejected at runtime anyway.
  - Lower underflow (user_min < exp_min) → ValueError, except the (user_min=1, exp_min=2) case which is just PyTorch's 0/1-specialization artifact (Dim(min=1) silently becomes 2 in ShapeEnv), this particular case emits a warning.                                                                                                                                                                                                       
  -  Subset (user range strictly inside exporter range) → engine profile is narrowed to the user's bounds (logged at INFO). The user's Input(min_shape, max_shape) is the contract for the TRT profile; a subset of the exporter's valid range is still correct, and a tighter profile lets TRT specialize more aggressively.                                                                                             
  - Dim.DYNAMIC (unbounded exporter upper) → no check, user fills the gap (the intended use).

### CODE REVIEW

**Comment by apbose** [comment]
CI index put cases failing. Looking at it.

**Comment by narendasan** [comment]
This PR is only for externally exported programs?

**Comment by apbose** [comment]
@narendasan yeah it addresses (torch_tensorrt.dynamo.compile(exported_program, ...)). The flow for torch_tensorrt.compile(model, inputs=..) should be unaffected, since in that case export bounds and input bounds should be same

**Review by narendasan — COMMENTED** [review]
You should use the user bounds in the narrow case

File: py/torch_tensorrt/dynamo/_compiler.py [hunk 1]
```diff
@@ -874,6 +876,139 @@ def _insert_complex_io_adapters(
...
+            if exp_max_unbounded:
+                # Dim.DYNAMIC: user fills the gap (intended use).
+                continue
+            try:
+                exp_min = int(exp_lower)
+                exp_max = int(exp_upper)
+            except (TypeError, ValueError):
+                continue
+            if user_min == exp_min and user_max == exp_max:
+                continue
```
Comment (narendasan) [inline py/torch_tensorrt/dynamo/_compiler.py:1024]: Can you explain what the logic here is?

File: py/torch_tensorrt/dynamo/_compiler.py [hunk 2]
```diff
@@ -874,6 +876,139 @@ def _insert_complex_io_adapters(
...
+                f"exporter({exp_min}, {exp_max})."
+            )
+            hint = (
+                f" Re-export with Dim('{expr}', min={user_min}, "
+                f"max={user_max}) or adjust Input to match."
+            )
+
+            if user_max > exp_max:
+                raise ValueError(
+                    f"{mismatch} Input.max_shape exceeds the exporter's max "
```
Comment (narendasan) [inline py/torch_tensorrt/dynamo/_compiler.py]: the exported program right?

File: py/torch_tensorrt/dynamo/_compiler.py [hunk 3]
```diff
@@ -874,6 +876,139 @@ def _insert_complex_io_adapters(
...
+                    f"({user_max} > {exp_max}); TRT will reject shapes above "
```
Comment (narendasan) [inline py/torch_tensorrt/dynamo/_compiler.py]: TRT will reject or pytorch will?

File: py/torch_tensorrt/dynamo/_compiler.py [hunk 4]
```diff
@@ -874,6 +876,139 @@ def _insert_complex_io_adapters(
...
+                    f"{exp_max} at runtime.{hint}"
+                )
+
+            if user_min < exp_min:
+                # 1->2 is the 0/1 specialization artifact, not a user error.
+                if user_min == 1 and exp_min == 2:
+                    logger.warning(
+                        "%s Input.min_shape=1 vs exporter min=2 is the "
```
Comment (narendasan) [inline py/torch_tensorrt/dynamo/_compiler.py]: I think we need to make sure these error messages are clear to follow from a user perspective

File: py/torch_tensorrt/dynamo/_compiler.py (hunk 1 above)
Comment (apbose) [reply py/torch_tensorrt/dynamo/_compiler.py:1024]: This block runs only when the exported program already bounds the symbol to a finite range (e.g. exported with Dim("batch", min=10, max=20)). The compiled engine's TRT optimization profile follows that exported range, so we compare the user's Input range against it here, at compile time, and handle four cases:
 - exported upper is unbounded (Dim.DYNAMIC) → user's Input fills the gap, nothing to validate (the continue just above);
- Input range == exported range → nothing to do;
- Input extends beyond the exported range (user_max > exp_max or user_min < exp_min) → those shapes can never run, so we raise now;
- Input strictly inside the exported range → safe to honor, we narrow the engine profile to the Input.
 The user_min == exp_min and user_max == exp_max check on this line is just the "exact match → no-op" early-out.

File: py/torch_tensorrt/dynamo/_compiler.py (hunk 2 above)
Comment (apbose) [reply py/torch_tensorrt/dynamo/_compiler.py]: yeah the exported program. will change the wording

File: py/torch_tensorrt/dynamo/_compiler.py (hunk 3 above)
Comment (apbose) [reply py/torch_tensorrt/dynamo/_compiler.py]: TRT will.  [10,20] + Input min_shape=2: shows
IExecutionContext::setInputShape: Error Code 3: API Usage Error
(Set dimension [2,8] for tensor x does not satisfy any optimization profiles.
 Valid range for profile 0: [10,8]..[20,8])
 through core/runtime/execute_engine.cpp:149.

**Comment by apbose** [comment]
failing tests- 
L0 dynamo core tests- RTX Linux- runtime cache failure
L0 dynamo core tests- Linux- same
L0 dynamo converter tests- RTX  Linux- cumsum operator
L0 torchscript tests- windows- test_e2e_behavior.py worker crashes,  test_get_layer_info Key Bindings missing
Python only runtime tests- wrong bounda

File: py/torch_tensorrt/dynamo/_compiler.py [hunk 5]
```diff
@@ -898,6 +900,149 @@ def _insert_complex_io_adapters(
...
+        sample_input = sample_by_name[node.target]
+        fake_val = node.meta.get("val")
+        if not isinstance(fake_val, torch.Tensor):
+            continue
+
+        min_shape = sample_input.shape["min_shape"]
+        max_shape = sample_input.shape["max_shape"]
+
+        for d, dim in enumerate(fake_val.size()):
+            if not isinstance(dim, torch.SymInt) or d >= len(min_shape):
```
Comment (narendasan) [inline py/torch_tensorrt/dynamo/_compiler.py]: in the case that its not SymInt shouldnt we check if they match the user dim?

File: py/torch_tensorrt/dynamo/_compiler.py [hunk 6]
```diff
@@ -929,6 +1074,12 @@ def compile_module(
     if sample_kwarg_inputs is None:
         sample_kwarg_inputs = {}
 
+    # Forwarded to the partitioner to fill Dim.DYNAMIC upper bounds.
+    # Read-only w.r.t. ShapeEnv so range_constraints survive save/re-export.
+    user_symbol_bounds = _build_user_symbol_bounds(
```
Comment (narendasan) [inline py/torch_tensorrt/dynamo/_compiler.py]: instead of trying to handle args, kwargs, why not operate on the flattened pytree inputs? If you really do need to do this work this early then you can always call in_spec to flatten the args, kwargs. Then the order is deterministic and you arent doing name matching

File: py/torch_tensorrt/dynamo/_compiler.py [hunk 7]
```diff
@@ -898,6 +900,149 @@ def _insert_complex_io_adapters(
...
+
+    Validates against finite exporter bounds: ``user_max > exp_max`` and
+    ``user_min < exp_min`` raise (TRT would reject those shapes at runtime);
+    a strict subset narrows the engine profile to the user's bounds (info
+    log only); the ``user_min=1, exp_min=2`` case warns -- it's PyTorch's
+    0/1 specialization artifact, not a user error.
+    """
+    placeholders = [n for n in gm.graph.nodes if n.op == "placeholder"]
+
+    sample_by_name: dict[str, Input] = {}
```
Comment (narendasan) [inline py/torch_tensorrt/dynamo/_compiler.py]: Like this code should go away with pytree flattening and you can just zip the placeholders

File: py/torch_tensorrt/dynamo/_compiler.py (hunk 5 above)
Comment (narendasan) [reply py/torch_tensorrt/dynamo/_compiler.py]: `d >= len(min_shape)`: in what case would we want this to be let through? the rank of the tensor cant change

File: py/torch_tensorrt/dynamo/_compiler.py [hunk 8]
```diff
@@ -898,6 +900,149 @@ def _insert_complex_io_adapters(
...
+        min_shape = sample_input.shape["min_shape"]
+        max_shape = sample_input.shape["max_shape"]
+
+        for d, dim in enumerate(fake_val.size()):
+            if not isinstance(dim, torch.SymInt) or d >= len(min_shape):
+                continue
+            expr = dim.node.expr
+            # Composite exprs (e.g. ``2*s0``) are recomputed by
+            # ``ShapeEnv.bound_sympy``; overriding them directly would lie.
+            if not isinstance(expr, sympy.Symbol):
```
Comment (narendasan) [inline py/torch_tensorrt/dynamo/_compiler.py:976]: You should log that this dynamic dimension is bounded by some other dynamic dimension

File: py/torch_tensorrt/dynamo/_compiler.py [hunk 9]
```diff
@@ -898,6 +901,164 @@ def _insert_complex_io_adapters(
...
+    ``user_min < exp_min`` raise (TRT would reject those shapes at runtime);
+    a strict subset narrows the engine profile to the user's bounds (info
+    log only); the ``user_min=1, exp_min=2`` case warns -- it's PyTorch's
+    0/1 specialization artifact, not a user error.
+    """
+    placeholders = [n for n in gm.graph.nodes if n.op == "placeholder"]
+
+    # Flatten args+kwargs in pytree order — guaranteed to match placeholder
+    # order by torch.export, so we can zip directly without name matching.
+    flat_inputs, _ = pytree.tree_flatten((list(sample_arg_inputs), sample_kwarg_inputs))
```
Comment (narendasan) [inline py/torch_tensorrt/dynamo/_compiler.py]: This should come from the exported program (i believe its called in_spec)

File: py/torch_tensorrt/dynamo/_compiler.py [hunk 10]
```diff
@@ -898,6 +901,164 @@ def _insert_complex_io_adapters(
...
+    """Map ``sympy.Symbol -> (min, max)`` from dynamic ``Input``s, used to
+    fill ``Dim.DYNAMIC`` upper bounds without mutating ``ShapeEnv``.
+
+    Validates against finite exporter bounds: ``user_max > exp_max`` and
+    ``user_min < exp_min`` raise (TRT would reject those shapes at runtime);
+    a strict subset narrows the engine profile to the user's bounds (info
+    log only); the ``user_min=1, exp_min=2`` case warns -- it's PyTorch's
+    0/1 specialization artifact, not a user error.
+    """
+    placeholders = [n for n in gm.graph.nodes if n.op == "placeholder"]
```
Comment (narendasan) [inline py/torch_tensorrt/dynamo/_compiler.py]: Same thing here, there should be a flat list of inputs to the graph

Comment (apbose) [reply py/torch_tensorrt/dynamo/_compiler.py]: we would need the placeholders for the faketensor values right?

Comment (apbose) [reply py/torch_tensorrt/dynamo/_compiler.py]: Pushed in the change to use in_spec. But I believe we still need  the placeholders, since in_spec would just give the ordering, but not the symbolic shape info?

Comment (narendasan) [reply py/torch_tensorrt/dynamo/_compiler.py]: `exported_program.graph_signature.user_inputs` I think this is the API to use because in theory with functionalized graphs (like the torch.compile graphs) you might have additional inputs that users have not provided right? We wouldnt have specs for those

Comment (apbose) [reply py/torch_tensorrt/dynamo/_compiler.py]: Updated the PR to use graph_signature.user_inputs for filtering placeholder nodes instead of iterating all placeholders. Need to rebase though

File: py/torch_tensorrt/dynamo/_compiler.py [hunk 11]
```diff
@@ -884,13 +888,174 @@ def _insert_complex_io_adapters(
...
+    fill ``Dim.DYNAMIC`` upper bounds without mutating ``ShapeEnv``.
+
+    Validates against finite exporter bounds: ``user_max > exp_max`` and
+    ``user_min < exp_min`` raise (TRT would reject those shapes at runtime);
+    a strict subset narrows the engine profile to the user's bounds (info
+    log only); the ``user_min=1, exp_min=2`` case warns -- it's PyTorch's
+    0/1 specialization artifact, not a user error.
+    """
+    name_to_node = {n.name: n for n in gm.graph.nodes if n.op == "placeholder"}
+    placeholders = [name_to_node[name] for name in graph_signature.user_inputs]
```
Comment (narendasan) [inline py/torch_tensorrt/dynamo/_compiler.py]: what user_inputs are not placeholders?

File: py/torch_tensorrt/dynamo/_compiler.py [hunk 12]
```diff
@@ -922,6 +1087,16 @@ def compile_module(
...
         )
+
+    # Forwarded to the partitioner to fill Dim.DYNAMIC upper bounds.
+    # Read-only w.r.t. ShapeEnv so range_constraints survive save/re-export.
+    if graph_signature is not None:
+        user_symbol_bounds = _build_user_symbol_bounds(
+            gm, sample_arg_inputs, sample_kwarg_inputs, graph_signature
+        )
+    else:
+        user_symbol_bounds = {}
```
Comment (narendasan) [inline py/torch_tensorrt/dynamo/_compiler.py]: Why would this case ever be triggered? What happens if it does?

File: py/torch_tensorrt/dynamo/_compiler.py (hunk 11 above)
Comment (apbose) [reply py/torch_tensorrt/dynamo/_compiler.py]: all user_inputs would be placeholders, but not all placeholders would be user_inputs. placeholders would also have the lifted params and buffers

File: py/torch_tensorrt/dynamo/_compiler.py (hunk 12 above)
Comment (apbose) [reply py/torch_tensorrt/dynamo/_compiler.py]: its a safety fallback if the compile_module is passed without the graph_signature. In that case we revert back to the behavior earlier. I can actually remove it and keep the graph_signature as not Optional. Both call sites in compile() calls it with the graph signature

File: py/torch_tensorrt/dynamo/_compiler.py [hunk 13]
```diff
@@ -884,6 +889,170 @@ def _insert_complex_io_adapters(
...
+    a strict subset narrows the engine profile to the user's bounds (info
+    log only); the ``user_min=1, exp_min=2`` case warns -- it's PyTorch's
+    0/1 specialization artifact, not a user error.
+    """
+    all_placeholders = [n for n in gm.graph.nodes if n.op == "placeholder"]
+    placeholders = [
+        p
+        for p, s in zip(all_placeholders, graph_signature.input_specs)
+        if s.kind == InputKind.USER_INPUT
+    ]
```
Comment (zewenli98) [inline py/torch_tensorrt/dynamo/_compiler.py]: Is the order of all_placeholders and graph_signature.input_specs guaranteed? 
If input_specs contains more entries (e.g., inserted lifted inputs), zip() will silently truncate to the shorter list and can misalign placeholders. 
If the graph contains placeholders in a different order, some user inputs will be skipped or matched to the wrong spec.

File: py/torch_tensorrt/dynamo/_compiler.py [hunk 14]
```diff
@@ -884,6 +889,170 @@ def _insert_complex_io_adapters(
...
+
+    in_spec = getattr(gm, "_in_spec", None)
+    assert in_spec is not None, "Exported graph module missing _in_spec"
+    flat_inputs = in_spec.flatten_up_to((list(sample_arg_inputs), sample_kwarg_inputs))
+
+    user_symbol_bounds: Dict[sympy.Symbol, Tuple[int, int]] = {}
+
+    for node, inp in zip(placeholders, flat_inputs):
```
Comment (zewenli98) [inline py/torch_tensorrt/dynamo/_compiler.py:944]: Are placeholders and flat_inputs matched as well?

File: py/torch_tensorrt/dynamo/_compiler.py (hunk 13 above)
Comment (apbose) [reply py/torch_tensorrt/dynamo/_compiler.py]: The ordering is guaranteed — input_specs and placeholder nodes are built together during torch.export(). Pytorch uses the same pattern in here https://github.com/pytorch/pytorch/blob/main/torch/export/exported_program.py#L1683-L1687

File: py/torch_tensorrt/dynamo/_compiler.py (hunk 14 above)
Comment (apbose) [reply py/torch_tensorrt/dynamo/_compiler.py:944]: In torch/export/_trace.py https://github.com/pytorch/pytorch/blob/main/torch/export/_trace.py#L1509
 _, original_in_spec = pytree.tree_flatten((args, kwargs))
This original_in_spec is what becomes gm._in_spec on the graph module (stored via ExportArtifact.in_spec). The placeholder nodes are created from the same pytree flattening of the same (args, kwargs) — so their order is determined by the same spec.

When we call in_spec.flatten_up_to((sample_arg_inputs, sample_kwarg_inputs)), we're reproducing that exact flattening order on the user's inputs. And USER_INPUT entries in input_specs are in the same placeholder order. Both come from the same source, so the zip should be correctly aligned.
